proposals/0008-swift-prometheus.md

```diff
@@ -0,0 +1,150 @@
+* Summary - Similar to a  *histogram* , a  *summary*  samples observations (usually things like request durations and response sizes). While it also provides a total count of observations and a sum of all observed values, it calculates configurable quantiles over a sliding time window.
+
+SwiftPrometheus provides fully featured implementations for all of them, including a thin wrapper around them for integration with `swift-metrics`.
+
+### API Layout
+Below section will lay out the public API of this package. For the internal APIs I would suggest you to read through the code on GitHub :smile:. This section is split up into two parts, using this library standalone, or using it integrated with the `swift-metrics` package. 
+
+#### Without swift-metrics
+To get started, initialise an instance of `PrometheusClient`
+```swift
+let myProm = PrometheusClient()
+```
+Once done, you can use the `create*` APIs to create any of the above described metric types.
+
+```swift
+// MetricLabels is a helper type used to add labeled metrics.
+struct MyCodable: MetricLabels {
+   var thing: String = "*"
+}
+
+// - Counter
+let counter = myProm.createCounter(forType: Int.self, named: "my_counter", helpText: "Just a counter", initialValue: 42, withLabelType: MyCodable.self)
+
+counter.inc() // Increment by one
+counter.inc(12) // Increment by a value
+counter.inc(12, MyCodable(thing: "test")) // Increment a labeled counter
+
+// - Gauge
+let gauge = myProm.createGauge(forType: Int.self, named: "my_gauge", helpText: "Just a gauge", initialValue: 42, withLabelType: MyCodable.self)
+gauge.inc() // Same APIs as Counter
+gauge.dec() // Same APIs as `inc()` but reversed.
+gauge.set(42) // Set the gauge to a specific value
+
+// - Histogram
+// Histograms use special labels, different than the Counter & Gauge
+struct HistogramLabel: HistogramLabels {
+   var le: String = ""
+   let route: String
+
+   init() {
+       self.route = "*"
+   }
+
+   init(_ route: String) {
+       self.route = route
+   }
+}
+
+let histogram = myProm.createHistogram(forType: Double.self, named: "my_histogram", helpText: "Just a histogram", labels: HistogramLabel.self)
+
+histogram.observe(123) // Observes a value
+
+// - Summary
+// Like Histograms, Summaries use different label types.
+struct SummaryLabel: SummaryLabels {
+   var quantile: String = ""
+   let route: String
+
+   init() {
+       self.route = "*"
+   }
+
+   init(_ route: String) {
+       self.route = route
+   }
+}
+
+let summary = myProm.createSummary(forType: Double.self, named: "my_summary", helpText: "Just a summary", labels: SummaryLabel.self)
+
+summary.observe(123) // Observes a value
+```
+Then, after you have some metric types, you can use `.collect()` on your `PromtheusClient` to get your Prometheus formatted string with all the data.
+For example, in a Vapor app:
+```swift
+router.get("/metrics") { req -> String in 
+    return myProm.collect()
+}
+```
+
+#### With swift-metrics
```

Comment:
i suggest putting this section before "Without swift-metrics" since the goal is for users to use the standard API unless they need to reach out to features not covered by it

Reply:
The reason I have the "Without" section first is because of the fact it allows for the more powerfull APIs. How I understood the role of `swift-metrics` was that it should be used by packages and libraries, but that end users should (if desired by the library) use the library directly.

If that's not the case, I'm open to changing this 😄

Reply:
the stated goal of the standard API is to encourage mobility between different backends solutions. prometheus is popular today, but another solution is likely to appear with time and replace it - the same way that prometheus replaced previous popular choices. as such we strongly encourage applications to use the standard API where possible and minimize reliance on solution-specific APIs

Reply:
I don't have strong feelings, but I do think if in general we're leaning toward using `swift-metrics` then that makes sense. I will say though, except for widely-used OSS projects, internally I'm likely to mostly use `SwiftPrometheus` directly to leverage its extra capabilities (like the help text associated with a metric) since [we're heavily invested](https://vimeo.com/274820706) there.

Reply:
@Yasumoto as mentioned above, the goal of the standard API is to encourage mobility between different backends solutions. as such, we expect libraries to *always* adopt the swift-metrics API, and applications to *prefer* using swift-metrics API unless they need features from the specific backend implementation eg `SwiftPrometheus`. my suggestion is to make this document more clearly reflect these design goals. its close, but still comes across as encouraging using the `SwiftPrometheus` API over the standard one, which can cause users trouble with code mobility

Reply:
I would like to second this, can we please re-order the "with/without" so that with swift metrics is first?

This does not mean that using this lib directly isn't good -- totally is, but we really need to put out first that this integrates with the metrics ecosystem (for libraries). And if users consciously _want to_ use this lib -- as you say @Yasumoto, you'd deliberately want to use the lib directly, but you'd really also want all libraries you use to use the metrics-api, rather than, say, statsd, since you want all libs to talk to your preferred backed: prometheus. :)

Reply:
Reordered them. Thanks for the input guys! 😄

Reply:
100% the right move, thanks folks! 🎉